indra/llcommon/llcallbacklist.h

```diff
@@ -27,53 +27,248 @@
-	//
-	typedef std::list< callback_pair_t >	callback_list_t; 
-	
-	LLCallbackList();
+	typedef boost::signals2::signal<void()> callback_list_t;
+	typedef callback_list_t::slot_type callable_t;
+	typedef boost::signals2::connection handle_t;
+	typedef boost::signals2::scoped_connection temp_handle_t;
+	typedef std::function<bool ()> bool_func_t;
+
 	~LLCallbackList();
 
-	void addFunction( callback_t func, void *data = NULL );			// register a callback, which will be called as func(data)
+	handle_t addFunction( callback_t func, void *data = NULL );		// register a callback, which will be called as func(data)
+	handle_t addFunction( const callable_t& func );
 	bool containsFunction( callback_t func, void *data = NULL );	// true if list already contains the function/data pair
 	bool deleteFunction( callback_t func, void *data = NULL );		// removes the first instance of this function/data pair from the list, false if not found
-	void callFunctions();														// calls all functions
+	void deleteFunction( const handle_t& handle );
+	void callFunctions();											// calls all functions
 	void deleteAllFunctions();
 
+	handle_t doOnIdleOneTime( const callable_t& func );
+	handle_t doOnIdleRepeating( const bool_func_t& func );
+	bool isRunning(const handle_t& handle) const { return handle.connected(); };
+
 	static void test();
 
 protected:
-
-	inline callback_list_t::iterator find(callback_t func, void *data);
-
 	callback_list_t	mCallbackList;
+
+	// "Additional specializations for std::pair and the standard container
+	// types, as well as utility functions to compose hashes are available in
+	// boost::hash."
+	// https://en.cppreference.com/w/cpp/utility/hash
+	typedef std::pair< callback_t,void* >	callback_pair_t;
+	typedef std::unordered_map<callback_pair_t, handle_t,
+		boost::hash<callback_pair_t>> lookup_table;
+	lookup_table mLookup;
 };
 
-typedef boost::function<void ()> nullary_func_t;
-typedef boost::function<bool ()> bool_func_t;
+/*-------------------- legacy names in global namespace --------------------*/
+#define gIdleCallbacks (LLCallbackList::instance())
+
+using nullary_func_t = LLCallbackList::callable_t;
+using bool_func_t    = LLCallbackList::bool_func_t;
 
 // Call a given callable once in idle loop.
-void doOnIdleOneTime(nullary_func_t callable);
+inline
+LLCallbackList::handle_t doOnIdleOneTime(nullary_func_t callable)
+{
+    return gIdleCallbacks.doOnIdleOneTime(callable);
+}
 
 // Repeatedly call a callable in idle loop until it returns true.
-void doOnIdleRepeating(bool_func_t callable);
+inline
+LLCallbackList::handle_t doOnIdleRepeating(bool_func_t callable)
+{
+    return gIdleCallbacks.doOnIdleRepeating(callable);
+}
+
+/*****************************************************************************
+*   LLLater: callbacks at some future time
+*****************************************************************************/
+class LLLater: public LLSingleton<LLLater>
```

Comment:
I'm really not a fan of this name.
For once: L-L-L-ater. How man L can we stuff in there? YES!

Also it is very unspecific. For the periodic calls it is close to wrong. As they are not strictly called later, but at internals.

Maybe ScheduleFunction? Or split the periodic and one shot functions into "InvokeLater" and "InvokePeriodically"?

(I'm bad at naming too!)